### PR TITLE
[DataGrid] Drop support of v4 below `4.12.0`

### PR DESCRIPTION
**Pull request by DanailH**
Found some errors being thrown in the console on some pages in the docs. This PR aims to fix those.

The main one is the deprecated `createMuiTheme` and since the docs use the latest version of the core this results in errors in the console.

The other one is an error related to having `key` being passed to the `GridToolbarFilterButton`.

---

###  Breaking changes

Drop support of v4 below `"@material-ui/core": "4.12.0"`

### CODE REVIEW

File: docs/src/pages/components/data-grid/style/StylingRowsGrid.tsx [hunk 1]
```diff
@@ -1,14 +1,14 @@
 import * as React from 'react';
 import { DataGrid } from '@material-ui/data-grid';
 import { useDemoData } from '@material-ui/x-grid-data-generator';
-import { createMuiTheme, darken, lighten, Theme } from '@material-ui/core/styles';
+import { createTheme, darken, lighten, Theme } from '@material-ui/core/styles';
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/style/StylingRowsGrid.tsx:4]: If we do this change, we need to change https://github.com/mui-org/material-ui-x/blob/30512081208da6b47922950206bc72cf18254c24/packages/grid/data-grid/package.json#L48

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx [hunk 2]
```diff
@@ -76,9 +76,9 @@ export const GridToolbarFilterButton = React.forwardRef<
       <div>
         {apiRef!.current.getLocaleText('toolbarFiltersTooltipActive')(counter)}
         <ul className={classes.list}>
-          {activeFilters.map((item) => ({
+          {activeFilters.map((item, index) => ({
             ...(lookup[item.columnField!] && (
-              <li key={item.id}>
+              <li key={index}>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:81]: +1 for reverting this change that Damien did recently.

File: docs/src/pages/components/data-grid/columns/ColumnTypesGrid.js [hunk 3]
```diff
@@ -35,32 +35,31 @@ const rows = [
   },
 ];
 
+const columns = [
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/columns/ColumnTypesGrid.js:38]: Why this change?

Comment (DanailH) [reply docs/src/pages/components/data-grid/columns/ColumnTypesGrid.js:38]: I'll rever that, its from another PR.

File: docs/src/pages/components/data-grid/style/StylingRowsGrid.tsx (hunk 1 above)
Comment (DanailH) [reply docs/src/pages/components/data-grid/style/StylingRowsGrid.tsx:4]: I just noticed the problem and since the docs use the latest version I thought it is safe to make the change.

Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/style/StylingRowsGrid.tsx:4]: Ok, then I would label this as a BC: drop support of v4 < 4.12.0, and update the peer dependency scope.

Comment (DanailH) [reply docs/src/pages/components/data-grid/style/StylingRowsGrid.tsx:4]: Updated

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx (hunk 2 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:81]: So keep that change right?

File: packages/grid/data-grid/package.json [hunk 4]
```diff
@@ -45,8 +45,8 @@
     "reselect": "^4.0.0"
   },
   "peerDependencies": {
-    "@material-ui/core": "^4.9.12 || ^5.0.0-beta.0",
-    "@material-ui/styles": "^4.9.10 || ^5.0.0-beta.0",
+    "@material-ui/core": "^4.12.0 || ^5.0.0-beta.0",
+    "@material-ui/styles": "^4.12.0 || ^5.0.0-beta.0",
```
Comment (m4theushw) [inline packages/grid/data-grid/package.json:49]: The lowest @material-ui/styles required by the core v4 is 4.12.0.

https://github.com/mui-org/material-ui/blob/53a1655143aa4ec36c29a6063ccdf89c48a74bfd/packages/material-ui/package.json#L53

```suggestion
    "@material-ui/styles": "^4.11.4 || ^5.0.0-beta.0",
```

**Comment by oliviertassinari** [comment]
It looks great. I think that we can keep the logic that supports < v4.12.0 as long as the overhead for us is low, it seems to be the case, so far.

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx (hunk 2 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:81]: Yeap
